src/Compilers/Core/Portable/Serialization/SerializationThreadPool.cs

```diff
@@ -0,0 +1,181 @@
+                enqueue((threadStart, state, tcs));
+
+                // Return the task.
+                return tcs.Task;
+
+                static void createThread()
+                {
+                    // Create a new background thread to run the work.
+                    var t = new Thread(() =>
+                    {
+                        // Repeatedly get the next item and invoke it, setting its TCS when we're done.
+                        // This will wait for up to the idle time before giving up and exiting.
+                        while (tryDequeue(out var item))
+                        {
+                            try
+                            {
+                                if (item.action is Func<object?, object?> callbackWithState)
+                                {
+                                    item.tcs.SetResult(callbackWithState(item.state));
+                                }
+                                else
+                                {
+                                    item.tcs.SetResult(((Func<object?>)item.action)());
+                                }
+                            }
+                            catch (OperationCanceledException ex)
+                            {
+                                item.tcs.TrySetCanceled(ex.CancellationToken);
+                            }
+                            catch (Exception ex)
+                            {
+                                item.tcs.TrySetException(ex);
+                            }
+                        }
+                    });
+                    t.IsBackground = true;
+                    t.Start();
+                }
+
+                static void enqueue((Delegate, object?, TaskCompletionSource<object?>) item)
+                {
+                    // Enqueue the work. If there are currently fewer threads waiting
+                    // for work than there are work items in the queue, create another
+                    // thread. This is a heuristic, in that we might end up creating
+                    // more threads than are truly needed, but this whole type is being
+                    // used to replace a previous solution where every work item created
+                    // its own thread, so this is an improvement regardless of any
+                    // such inefficiencies.
+                    lock (s_queue)
+                    {
+                        s_queue.Enqueue(item);
+
+                        if (s_queue.Count <= s_availableThreads)
+                        {
+                            Monitor.Pulse(s_queue);
+                            return;
+                        }
+                    }
+
+                    // No thread was currently available.  Create one.
+                    createThread();
+                }
+
+                static bool tryDequeue(out (Delegate action, object? state, TaskCompletionSource<object?> tcs) item)
```

Comment:
Consider including these parameter names in similar tuples in this file.

Reply:
➡️ Now updated